### DISCUSSION TITLE
Menu border

### DISCUSSION THREAD

**Discussion by christoph-heinrich** [tomasklaen/uosc#732] [Ideas]
There has recently been [that change to the curtain](https://github.com/tomasklaen/uosc/commit/2ddc4d11f096442a7973800c7925cf94b0449a27), which I really don't like and thus configured things back to how they were. However that commit did a address a real problem that I'm somewhat bothered by, so what if instead we draw a faint border around the menu to separate it from the background.

Here is a test with a 1px border the same color and opacity as the title separator, on top of the menu background.
<details>
<summary>diff</summary>

```diff
diff --git a/scripts/uosc/elements/Menu.lua b/scripts/uosc/elements/Menu.lua
index bf0a609..0899216 100644
--- a/scripts/uosc/elements/Menu.lua
+++ b/scripts/uosc/elements/Menu.lua
@@ -1113,6 +1113,10 @@ function Menu:render()
 		ass:rect(menu_rect.ax, menu_rect.ay, menu_rect.bx, menu_rect.by, {
 			color = bg, opacity = menu_opacity * config.opacity.menu, radius = state.radius + self.padding,
 		})
+		ass:rect(menu_rect.ax + 1, menu_rect.ay + 1, menu_rect.bx - 1, menu_rect.by - 1, {
+			color = fg, opacity = {main = 0, border = menu_opacity * 0.2}, radius = state.radius + self.padding - 1,
+			border_color = fg, border = 1
+		})
 
 		if is_parent then
 			cursor:zone('primary_down', menu_rect, self:create_action(function() self:slide_in_menu(menu, x) end))
@@ -1253,7 +1257,7 @@ function Menu:render()
 			end
 
 			-- Bottom border
-			ass:rect(ax, rect.by - self.separator_size, bx, rect.by, {color = fg, opacity = menu_opacity * 0.2})
+			ass:rect(ax + 1, rect.by - self.separator_size, bx - 1, rect.by, {color = fg, opacity = menu_opacity * 0.2})
 
 			-- Do nothing when user clicks title
 			if is_current then
```
</details>

```
color=curtain=000000
opacity=menu=0.7,curtain=0.5
```
![image](https://github.com/tomasklaen/uosc/assets/8932183/0f0868dc-0c6f-4793-ba72-0c7b14e6fddc)
![image](https://github.com/tomasklaen/uosc/assets/8932183/37fa72c2-226f-4735-b547-586446a27a25)

At first I also had a version where I simply added a border, but when it's not on a the dark background from the menu it looks like a halo on a light frame. But at the same time the edge looks more defined, the other version looks a little blurry on a light frame... not sure which is preferable.
![image](https://github.com/tomasklaen/uosc/assets/8932183/e9b770a2-5f74-4ccc-adb0-85c66fdef5e2)
![image](https://github.com/tomasklaen/uosc/assets/8932183/fb8af171-aba5-4fb1-aaf7-7abc5ea68ad7)

While adding those last two I had the idea of offsetting the border in a bit, so that the dark background is actually around it, then it wouldn't look blurry on a light frame.
![image](https://github.com/tomasklaen/uosc/assets/8932183/f3a143ca-aacf-42df-9f29-e77df62b5d9c)
![image](https://github.com/tomasklaen/uosc/assets/8932183/005eee99-81f0-41de-80f1-336b26859207)

**Comment by tomasklaen** [comment]
I made the curtain change precisely so that we don't have to add borders :( I hate borders. I always try to only have edges, and never borders. They always look bad. As you pointed out, border **on** the background will make it look blurry against lighter backgrounds, and border **around** the background will give it an ugly halo. The offset border is starting to go into faux 3D design, which is out of place and I don't want the UI looking like that.

> not sure which is preferable.

The best compromise is honestly having a slightly different background or curtain color, than there's never going to be a blend of the two, and you can always see a defined edge.

I'm not against changing it so that the default background is `111111` and curtain `000000`. It actually makes more sense that way.

**Reply by tomasklaen** [reply]
> I'm not against changing it so that the default background is 111111 and curtain 000000. It actually makes more sense that way.

I've actually tried it and imo it looks worse than current default. But defaults don't matter much, everyone can change them to suit their taste.

**Reply by christoph-heinrich** [reply]
I'm not a fan of that whole gray thing in general and would rather keep it black.

Also that doesn't really solve the problem and instead makes it less likely to happen, but there will still be some frame color that together with the curtain results in the same color as the menu background. The only way to avoid that would be an opaque curtain or some sort of border around the menu.

The offset border doesn't look 3D to me at all.

**Reply by dyphire** [reply]
Well, color settings are subjective. However, I also don't really like borders, as they are a bit bad at 4k resolution rendering. The offset border does seem a bit pseudo 3D on a light frame to me.

**Reply by christoph-heinrich** [reply]
Weird that it seems 3D to both of you but not to me. I can't think of any other solution, so I guess we'll leave it as is. It's not like that reduces the usability, so it's not that important.

**Comment by dyphire** [comment]
I think the problem here is that we use pure black as the background, which leads to insufficient differentiation. 
The color settings mentioned [there](https://github.com/tomasklaen/uosc/issues/733#issuecomment-1774513653) have been good improvements.
```
color=foreground=FFFBFE,foreground_text=1C1B1F,background=1C1B1F,background_text=FFFBFE
```
